build_zng.rs

```diff
@@ -1,58 +1,691 @@
-use std::env;
+use std::{
+    collections::{BTreeMap, BTreeSet},
+    env, fs,
+    hash::{Hash as _, Hasher as _},
+    io::Write as _,
+    path::{Path, PathBuf},
+};
+
+fn append(cfg: &mut cc::Build, root: Option<&str>, files: impl IntoIterator<Item = &'static str>) {
+    let root = root.unwrap_or("");
+    cfg.files(
+        files
+            .into_iter()
+            .map(|fname| format!("src/zlib-ng/{root}{fname}.c")),
+    );
+}
+
+/// Replicate the behavior of cmake/make/configure of stripping out the
+/// @ZLIB_SYMBOL_PREFIX@ since we don't want or need it
+fn strip_symbol_prefix(input: &Path, output: &Path, mut on_line: impl FnMut(&str)) {
+    let contents = fs::read_to_string(input).unwrap();
+    let mut h = fs::File::create(output).expect("failed to create zlib include");
+
+    use std::io::IoSlice;
+    let mut write = |bufs: &[IoSlice]| {
+        // write_all_vectored is unstable
+        for buf in bufs {
+            h.write_all(&buf).unwrap();
+        }
+    };
+
+    for line in contents.lines() {
+        if let Some((begin, end)) = line.split_once("@ZLIB_SYMBOL_PREFIX@") {
+            write(&[
+                IoSlice::new(begin.as_bytes()),
+                IoSlice::new(end.as_bytes()),
+                IoSlice::new(b"\n"),
+            ]);
+        } else {
+            write(&[IoSlice::new(line.as_bytes()), IoSlice::new(b"\n")]);
+        }
+
+        on_line(line);
+    }
+}
+
+#[derive(Default)]
+struct AppendState {
+    flags: BTreeSet<&'static str>,
+    defines: BTreeSet<&'static str>,
+    files: BTreeSet<&'static str>,
+}
+
+#[derive(Debug, Hash, Copy, Clone)]
+struct TargetFeature {
+    check: &'static str,
+    msvc_flags: &'static [&'static str],
+    flags: &'static [&'static str],
+    defines: &'static [&'static str],
+    files: &'static [&'static str],
+}
+
+struct Ctx<'ctx> {
+    cfg: &'ctx mut cc::Build,
+    root: Option<&'ctx str>,
+    enabled: Option<&'ctx BTreeSet<&'ctx str>>,
+    msvc: bool,
+    append: AppendState,
+    cache: Option<BTreeMap<u64, bool>>,
```

Comment:
Why not use:

```suggestion
    cache: Option<BTreeMap<TargetFeature, bool>>,
```

Why do you manually calculate the hash instead of using hashmap?

Hash can collide and without storing `TargetFeature`, you can't perform an equality check to make sure it isn't a , which HashMap already does the job for you.

I also find no reason to use HashMap over BTreeMap since 8MO there aren't enough elements for HashMap to outperform BTreeMap, it only slows down compilation since HashMap is more complex than BTreeMap and you would have to `derive(Hash)`

Reply:
I don't understand the complaint here, you are both saying to use hashmap and not use hashmap?

Reply:
I'm saying that using hasher directly with a BTreeMap is wrong and you should use HashMap instead of hasher + BTreeMap.

However given the amount of data I don't think HashMap is necessary and BTreeMap might be faster and has less compilation overhead.